src/html/XHPHelpers.php

```diff
@@ -187,4 +134,44 @@ public function requireUniqueID(): string {
     }
   }
 
+  final public function transferAttributesToRenderedRoot(
+    :x:composable-element $root,
+  ): void {
+    if (:xhp::$ENABLE_VALIDATION && $root instanceof :x:element) {
+      if (!($root instanceof HasXHPHelpers)) {
+        throw new XHPClassException(
+          $this,
+          'render() must return an object using the XHPHelpers trait.'
+        );
+      }
+
+      $rootID = $root->getAttribute('id') ?: null;
+      $thisID = $this->getAttribute('id') ?: null;
+
+      if ($rootID && $thisID && $rootID != $thisID) {
+        throw new XHPException(
+          'ID Collision. '.(:xhp::class2element(self::class)).' has an ID '.
+          'of "'.$thisID.'" but it renders into a(n) '.
+          (:xhp::class2element(get_class($root))).
+          ' which has an ID of "'.$rootID.'". The latter will get '.
+          'overwritten (most often unexpectedly). If you are intending for '.
+          'this behavior consider calling $this->removeAttribute(\'id\') '.
+          'before returning your node from compose().'
+        );
+      }
+    }
+    assert($root instanceof HasXHPHelpers);
+
+    $attributes = $this->getAttributes();
+
+    // We want to append classes to the root node, instead of replace them,
+    // so do this attribute manually and then remove it.
+    if (array_key_exists('class', $attributes) && $attributes['class']) {
```

Comment:
There should be something on the class to denote that a set of attributes should be appended instead of replaced. For normal HTML that's just class, but there can be others. For instance, Javelin needs `data-sigil`. What about putting a function in this trait that is like

```
protected function getAttributesThatAppendValuesOnTransfer(): ImmSet<string> {
  return ImmSet {'string'};
}
```

And you can override it individual classes or create your own trait that extends this to add your special-needs attributes.